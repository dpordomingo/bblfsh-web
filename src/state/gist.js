import * as history from 'services/history';
import * as api from 'services/api';
import { set as codeSet } from './code';
import { reset as examplesReset } from './examples';
import { add as errorsAdd } from './errors';

const initialState = {
  url: '',
  gist: '',
  isValid: false,
};

export const SET_URL = 'bblfsh/gist/SET_URL';
export const SET = 'bblfsh/gist/SET';

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_URL:
      const parts = action.url.match(gistRegexp);
      const gist = parts === null ? '' : parts[1];

      return {
        ...reducer(state, {
          type: SET,
          gist,
        }),
        url: action.url,
      };
    case SET:
      return {
        ...state,
        url: `https://gist.githubusercontent.com/${action.gist}`,
        gist: action.gist,
        isValid: !!action.gist,
      };
    default:
      return state;
  }
};

const gistRegexp = new RegExp(
  '^\\s*(?:https?://)?gist.githubusercontent.com/(\\S+\\s*$)'
);

export const setURL = url => {
  return {
    type: SET_URL,
    url,
  };
};

export const set = gist => {
  return {
    type: SET,
    gist,
  };
};

export const load = () => (dispatch, getState) => {
  const { gist, isValid } = getState().gist;
  if (!isValid) {
    return Promise.reject();
  }

  const gistParts = gist.split('/');
  const filename = gistParts[gistParts.length - 1];

  dispatch(examplesReset());
  // cleanup code
  dispatch(codeSet('', null));

  return api
    .getGist(gist)
    .then(content => {
      dispatch(codeSet(filename, content));
      // url side effect
      history.setGist(gist);
    })
    .catch(errors => {
      dispatch(errorsAdd(errors));
      return Promise.reject();
    });
};
