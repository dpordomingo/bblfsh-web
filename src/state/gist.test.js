import { reducer, initialState, SET_URL, SET } from './gist';

describe('gist/reducer', () => {
  it('SET_URL', () => {
    expect(
      reducer(initialState, {
        type: SET_URL,
        url: 'https://gist.githubusercontent.com/gist1',
      })
    ).toMatchSnapshot();
    expect(
      reducer(initialState, {
        type: SET_URL,
        url: 'invalid',
      })
    ).toMatchSnapshot();
  });

  it('SET', () => {
    expect(
      reducer(initialState, {
        type: SET,
        gist: 'gist1',
      })
    ).toMatchSnapshot();
  });
});
