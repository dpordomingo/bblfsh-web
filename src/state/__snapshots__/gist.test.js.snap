// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gist/reducer SET 1`] = `
Object {
  "gist": "gist1",
  "isValid": true,
  "url": "https://gist.githubusercontent.com/gist1",
}
`;

exports[`gist/reducer SET_URL 1`] = `
Object {
  "gist": "gist1",
  "isValid": true,
  "url": "https://gist.githubusercontent.com/gist1",
}
`;

exports[`gist/reducer SET_URL 2`] = `
Object {
  "gist": "",
  "isValid": false,
  "url": "invalid",
}
`;
